### PR TITLE
feat(unstable): add more options to Deno.createHttpClient

### LINKED ISSUE #11056: [extention/fetch] Allow reqwest HTTPClient configuration
Currently configuration for underlying reqwest httpclient is concealed, only  [TLS certificate](https://github.com/denoland/deno/blob/7b559ee89ef2dd4be6a45af6ae1e3e09c76fec01/cli/dts/lib.deno.unstable.d.ts#L1078) being exposed. Creating shared HTTP client uses some options predefined both [by Deno](https://github.com/denoland/deno/blob/main/extensions/fetch/lib.rs#L516) and [reqwest itself](https://docs.rs/reqwest/0.11.3/src/reqwest/async_impl/client.rs.html#126).

Controlling HTTP client configuration is very important for server-side applications. For instance, inability to configure tcp connection pool for http/1.1 is a deal breaker.

The complication I am seeing is agreeing on the API. I'm not sure if Deno wishes to use reqwest forever or move to something else in the future, therefore configuration options relying on reqwest too much might be dangerous. In that case I would propose avoid the risk of leaky abstraction and expose reqwest explicitly with [its options](https://docs.rs/reqwest/0.11.3/reqwest/struct.ClientBuilder.html):
```typescript
export interface CreateHttpClientOptions {
    clientProvider: "reqwest" | "other implementation"

    /** A certificate authority to use when validating TLS certificates. Certificate data must be PEM encoded.
     */
    caData?: string;
  }

export interface CreateReqwestHttpClientOptions extends CreateHttpClientOptions {
    clientProvider: "reqwest"
    poolIdleTimeout: number
    poolMaxIdlePerHost: number
    ...
}
```

What do you think? Do you have any other plans/ideas?

Thank you!

### LINKED ISSUE #13636: APIs to facilitate connection pools
I am unsure if [`Deno.connect`](https://doc.deno.land/deno/stable/~/Deno.connect) requires clients to implement connection pooling, because the current API surface that `Deno.Conn` exposes makes it tricky to implement one (though, [folks indeed have impl poolers atop it using `try..catch` semantics](https://github.com/denodrivers/postgres/blob/b68d3ebfd1a5805050f12e74ddaa4df3185bb86f/client.ts#L486)).

In comparing with NodeJS `net`, I believe equivalent of these two APIs are needed in [`Deno.Conn`](https://doc.deno.land/deno/stable/~/Deno.Conn) (TCP):

1. [`socket.readyState`](https://nodejs.org/api/net.html#socketreadystate) tells if conn is open or half-open.
2. [`socket.destroyed`](https://nodejs.org/api/net.html#socketdestroyed) tells if conn is open at all.
3. An adjustable [`socket.keepalive`](https://nodejs.org/api/net.html#socketsetkeepaliveenable-initialdelay).

([Here's how we use those APIs to determine health of a tcp-conn](https://github.com/serverless-dns/serverless-dns/blob/688c3110f6/src/core/dns/conns.js#L20-L22)).

NodeJS `net` also exposes [`socket.resume`](https://nodejs.org/api/net.html#socketresume) and [`socket.pause`](https://nodejs.org/api/net.html#socketpause), though those are not really required to build a pooler.

From NodeJS `dgram`, for _connected_ UDP sockets, [`Event.close`](https://nodejs.org/api/dgram.html#event-close) and `Event.error` come in handy to remove failed conns preemptively from the pool.

([Here's how we use those events to evict udp-conns from the pool](https://github.com/serverless-dns/serverless-dns/blob/688c3110f6/src/core/dns/conns.js#L152-L153)).

Of course, if Deno doesn't require client-code to pool (TCP/UDP) conns at all (like for HTTP), then that's even more awesome. [From this github discussion](https://github.com/tokio-rs/tokio/issues/1079), it is confirmed that `tokio` doesn't pool conns, but that [`tower`](https://github.com/tower-rs/tower) does... but I am not fluent in Rust to deduce anything from Deno's use of `tower`.

cc: @bartlomieju

### PR DESCRIPTION
**Pull request by crowlKats**
Closes #11056
Closes #13636

### CODE REVIEW

**Comment by crowlKats** [comment]
still need to add tests; unsure exactly how to test some of these though

File: ext/fetch/lib.rs [hunk 1]
```diff
@@ -650,13 +677,17 @@ where
 
 /// Create new instance of async reqwest::Client. This client supports
 /// proxies and doesn't follow redirects.
+#[allow(clippy::too_many_arguments)]
```
Comment (aapoalas) [inline ext/fetch/lib.rs:680]: Maybe time to create some optional params struct? Though I'm guessing that might've existed at some point and been removed for performance reasons?

**Review by aapoalas — APPROVED** [review]
LGTM. Tests would indeed be good: There is this PR https://github.com/denoland/deno/pull/16210 that has a use-case for `pool_max_idle_per_host` at least. That may serve as some idea for testing.

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 2]
```diff
@@ -748,6 +748,13 @@ declare namespace Deno {
     certChain?: string;
     /** PEM formatted (RSA or PKCS8) private key of client certificate. */
     privateKey?: string;
+    /** Sets the maximum idle connection per host allowed in the pool. */
+    poolMaxIdlePerHost?: number;
+    /** Set an optional timeout for idle sockets being kept-alive.
+     * Set to false to disable the timeout. */
+    poolIdleTimeout?: number | false;
+    /** Specify whether only wanting HTTP1 or HTTP2 */
```
[the remark this answers is not in the document]
Comment (crowlKats) [reply cli/tsc/dts/lib.deno.unstable.d.ts:756]: sure

[the remark this answers is not in the document]
Comment (crowlKats) [reply cli/tsc/dts/lib.deno.unstable.d.ts:756]:
```suggestion
    /** Specify whether only wanting HTTP/1.1 or HTTP/2 */
```

File: ext/fetch/lib.rs (hunk 1 above)
Comment (crowlKats) [reply ext/fetch/lib.rs:680]: Never has existed; will add one

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 3]
```diff
@@ -821,6 +821,13 @@ declare namespace Deno {
     certChain?: string;
     /** PEM formatted (RSA or PKCS8) private key of client certificate. */
     privateKey?: string;
+    /** Sets the maximum idle connection per host allowed in the pool. */
+    poolMaxIdlePerHost?: number;
+    /** Set an optional timeout for idle sockets being kept-alive.
+     * Set to false to disable the timeout. */
+    poolIdleTimeout?: number | false;
+    /** Specify whether only wanting HTTP/1.1 or HTTP/2 */
+    only?: "http1" | "http2";
```
Comment (lucacasonato) [inline cli/tsc/dts/lib.deno.unstable.d.ts:830]: Instead of an only option, each of HTTP 1 and http 2 should be a Boolean

File: ext/fetch/lib.rs [hunk 4]
```diff
@@ -699,6 +763,24 @@ pub fn create_http_client(
...
+      pool_idle_timeout.map(std::time::Duration::from_millis),
+    );
+  }
+
+  match (options.http1, options.http2) {
+    (true, false) => builder = builder.http1_only(),
+    (false, true) => builder = builder.http2_prior_knowledge(),
+    (true, true) => {}
+    (false, false) => {
+      return Err(type_error("At least http1 or http2 need to be enabled"))
```
Comment (bartlomieju) [inline ext/fetch/lib.rs:781]:
```suggestion
      return Err(type_error("Either `http1` or `http2` needs to be true"))
```

File: test_util/src/lib.rs [hunk 5]
```diff
@@ -1485,6 +1489,42 @@ async fn wrap_https_h2_only_server() {
...
+    if main_server_http.await.is_err() {
+      continue;
+    }
+  }
+}
+
+async fn wrap_https_h2_only_server() {
+  let main_server_http_addr = SocketAddr::from(([127, 0, 0, 1], H2_ONLY_PORT));
+
+  loop {
```
Comment (bartlomieju) [inline test_util/src/lib.rs:1513]: This loop can probably be gone

File: test_util/src/lib.rs [hunk 6]
```diff
@@ -1485,6 +1489,42 @@ async fn wrap_https_h2_only_server() {
   }
 }
 
+async fn wrap_https_h1_only_server() {
+  let main_server_http_addr = SocketAddr::from(([127, 0, 0, 1], H1_ONLY_PORT));
+
+  loop {
```
Comment (bartlomieju) [inline test_util/src/lib.rs:1495]: Ditto

File: cli/file_fetcher.rs [hunk 7]
```diff
@@ -1746,7 +1747,7 @@ mod tests {
 
   fn create_test_client() -> HttpClient {
     HttpClient::from_client(
-      create_http_client("test_client", None, vec![], None, None, None)
+      create_http_client("test_client", CreateHttpClientOptions::default())
```
Comment (bartlomieju) [inline cli/file_fetcher.rs:1750]: Good cleanup

**Comment by bartlomieju** [comment]
LGTM

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 8]
```diff
@@ -821,6 +821,15 @@ declare namespace Deno {
     certChain?: string;
     /** PEM formatted (RSA or PKCS8) private key of client certificate. */
     privateKey?: string;
+    /** Sets the maximum numer of idle connections per host allowed in the pool. */
+    poolMaxIdlePerHost?: number;
+    /** Set an optional timeout for idle sockets being kept-alive.
+     * Set to false to disable the timeout. */
+    poolIdleTimeout?: number | false;
+    /** Whether HTTP/1.1 is allowed or not. */
```
[the remark this answers is not in the document]
Comment (crowlKats) [reply cli/tsc/dts/lib.deno.unstable.d.ts:829]: > Should either say it defaults to true (but that is confusing)

How so?

[the remark this answers is not in the document]
Comment (lucacasonato) [reply cli/tsc/dts/lib.deno.unstable.d.ts:829]: Because undefined casts to false, not true. http1: undefined should therefore cast to http1: false

**Pull request merged**